### PR TITLE
Fix np.interp for np1.17 nan handling

### PR DESCRIPTION
**Pull request by sklam**
with PEP8 fixes.

### CODE REVIEW

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -1985,92 +2071,98 @@ def np_interp_impl_inner(x, xp, fp, dtype):
...
-                dres.flat[i] = lval
-            elif x_val > xp_val:
-                dres.flat[i] = rval
-            else:
-                dres.flat[i] = fp_val
-
-    else:
-        j = 0
+def np_interp_impl_inner_factory(np117_nan_handling):
+    @register_jitable
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: By having the decorator in the closure, does this lead to the potential to create a lot of instances of this? Do named instances at the module level help? i.e.

```python
_np_interp_post_np117 = register_jitable(_np_interp_impl_inner_factory)(True)
_np_interp_pre_np117 = register_jitable(_np_interp_impl_inner_factory)(False)
```

File: numba/targets/arraymath.py [hunk 2]
```diff
@@ -1803,103 +1857,135 @@ def np_interp_impl_complex_fp_inner(x, xp, fp, dtype):
...
+                else:
+                    if slopes.size:
+                        slope = slopes[j]
+                    else:
+                        inv_dx = 1 / (dx[j + 1] - dx[j])
+                        real = (dy[j + 1].real - dy[j].real) * inv_dx
+                        imag = (dy[j + 1].imag - dy[j].imag) * inv_dx
+                        slope = real + 1j * imag
+
+                    # The following branches is to mimick the behavior of
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]:
```suggestion
                    # The following branches mimic the behavior of
```

**Comment by stuartarchibald** [comment]
I also tested this manually against NumPy 1.17, needs #4357 merging in to make all the tests pass but they do pass.

File: numba/targets/arraymath.py (hunk 1 above)
Comment (sklam) [reply numba/targets/arraymath.py]: It wouldn't generate a lot of the instances due to the current in-memory cache of the `overload`.  But, i will change it as suggested for future-proof.

**Comment by seibert** [comment]
tiny flake8 fail on latest:
```
numba/targets/arraymath.py:2181:5: E303 too many blank lines (2)
```

**Comment by seibert** [comment]
Aside from a conda 4.7 related glitch finding NumPy 1.10, this has passed CI now.  Ready to merge.

**Pull request merged**